drivers/video/video_stm32_dcmi.c

```diff
@@ -0,0 +1,481 @@
+		LOG_ERR("DCMI DMA Init failed");
+		return -EIO;
+	}
+	LOG_INF("DCMI with DMA transfer");
+
+	return 0;
+}
+
+static int stm32_dcmi_enable_clock(const struct device *dev)
+{
+	const struct video_stm32_dcmi_config *config = dev->config;
+	const struct device *dcmi_clock = DEVICE_DT_GET(STM32_CLOCK_CONTROL_NODE);
+	int err;
+
+	if (!device_is_ready(dcmi_clock)) {
+		LOG_ERR("clock control device not ready");
+		return -ENODEV;
+	}
+
+	/* Turn on DCMI peripheral clock */
+	err = clock_control_on(dcmi_clock, (clock_control_subsys_t *) &config->pclken);
+	if (err < 0) {
+		LOG_ERR("Failed to enable DCMI clock");
+		LOG_ERR("Error code %d: %s", err, strerror(-err));
+		return err;
+	}
+
+	return 0;
+}
+
+static int video_stm32_dcmi_set_fmt(const struct device *dev,
+				  enum video_endpoint_id ep,
+				  struct video_format *fmt)
+{
+	const struct video_stm32_dcmi_config *config = dev->config;
+	struct video_stm32_dcmi_data *data = dev->data;
+
+	if (ep != VIDEO_EP_OUT) {
+		return -EINVAL;
+	}
+
+	data->pixel_format = fmt->pixelformat;
+	data->pitch = fmt->pitch;
+	data->height = fmt->height;
+	data->width = fmt->width;
+
+	if (config->sensor_dev && video_set_format(config->sensor_dev, ep, fmt)) {
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int video_stm32_dcmi_get_fmt(const struct device *dev,
+				  enum video_endpoint_id ep,
+				  struct video_format *fmt)
+{
+	struct video_stm32_dcmi_data *data = dev->data;
+	const struct video_stm32_dcmi_config *config = dev->config;
+
+	if (fmt == NULL || ep != VIDEO_EP_OUT) {
+		return -EINVAL;
+	}
+
+	if (config->sensor_dev && !video_get_format(config->sensor_dev, ep, fmt)) {
+		/* align DCMI with sensor fmt */
+		return video_stm32_dcmi_set_fmt(dev, ep, fmt);
+	}
+
+	fmt->pixelformat = data->pixel_format;
+	fmt->height = data->height;
+	fmt->width = data->width;
+	fmt->pitch = data->pitch;
+
+	return 0;
+}
+
+static int video_stm32_dcmi_stream_start(const struct device *dev)
+{
+	LOG_WRN("Start stream capture");
+
+	const struct video_stm32_dcmi_config *config = dev->config;
+
+	if (config->sensor_dev && video_stream_start(config->sensor_dev)) {
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int video_stm32_dcmi_stream_stop(const struct device *dev)
+{
+	LOG_WRN("Stop stream capture");
+
+	struct video_stm32_dcmi_data *data = dev->data;
+	const struct video_stm32_dcmi_config *config = dev->config;
+	int err;
+
+	if (config->sensor_dev && video_stream_stop(config->sensor_dev)) {
+		return -EIO;
+	}
+
+	err = HAL_DCMI_Stop(&data->hdcmi);
+	if (err != HAL_OK) {
+		LOG_ERR("Failed to stop DCMI");
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int video_stm32_dcmi_enqueue(const struct device *dev,
+				  enum video_endpoint_id ep,
+				  struct video_buffer *vbuf)
+{
+	struct video_stm32_dcmi_data *data = dev->data;
+
+	if (ep != VIDEO_EP_OUT) {
+		return -EINVAL;
+	}
+
+	vbuf->bytesused = data->pitch * data->height;
+
+	int err = HAL_DCMI_Start_DMA(&data->hdcmi, DCMI_MODE_SNAPSHOT,
```

Comment:
Any plan to support continuous capture?

Reply:
I don't have intention for this version. Maybe in a future as an improvement. 

Both sample codes for video capture are based on performing an enqueue command before dequeue. In this way, I believe that a snapshot strategy is enough.

Reply:
what If i we enqueue two buffers? should'nt we start DMA transfer only for the first one? and restart internally on frame done?

Reply:
The function `HAL_DCMI_Start_DMA` can be called only on time. If this function is called again, the `HAL_ERROR` is returned. 

In the ST sample code, only one video buffer is used. Below is described the workflow in those samples:
1. Start the DCMI via HAL_DCMI_Start_DMA in DCMI_MODE_CONTINUOUS.
2. When the HAL_DCMI_FrameEventCallback is trigger, call the HAL_DCMI_Suspend.
3. Retrieve the video buffer.
4. Resume the DCMI via HAL_DCMI_Resume.
5. Repeat item 2.

I can try to reproduce that workflow, but I believe that the result will be the same. Please, let me know if I missed something.

Reply:
I believe it's more sensible to retain the existing workflow, since the video buffer is passed through the enqueue function.

Reply:
I created this sample version. It'll work if you enqueue just one buffer. Link to the sample code https://github.com/zephyrproject-rtos/zephyr/pull/71463. 

Next week I'll try to look more into DCMI ST API to check if I can do better.

Reply:
Hi, @loicpoulain.

I changed the driver to work as continuous capture and with more than one video framebuffer.

Reply:
Since HAL_DCMI_Start_DMA can only work with one buffer, I created a dedicated buffer to store the frame data. This is the current workflow:
1. application enqueue one or more buffers.
2. call start capture function. The DCMI driver creates a buffer to store the frame data and starts in continuous mode.
3. when the interrupt occurs, call DCMI suspend function
4. If a video buffer is available, copy the frame data from the DCMI buffer to the video buffer.
5. resume the DCMI capture. 
6. go to item 2.

This driver was tested with this sample code https://github.com/zephyrproject-rtos/zephyr/pull/71463